### ISSUE TITLE
Number of rows in each chunk should depend on baseline length

### ISSUE THREAD

**Issue by bennahugo** [ska-sa/xova#25]
* xova version: 8be537fd19a168b5e673eabf003d5befea32090b, 
* africanus version: 55edd8cae11d51cb8c9a935c42ce6fa9e8a074a7 (manually kludged to work with py3.6.9)
* Python version: 3.6.9

Making some notes here before I forget what is wrong:

This here has been the source of 2 days worth of heartache - why is this fixed to a constant value for all arrays????
```
 94         p.add_argument("-rc", "--row-chunks", type=int, default=10000,
 95                        help="Number of rows averaged together "
 96                             "as part of a single, independent, "
 97                             "averaging operation."
```
The number of rows must, at absolute minimum, be nbl * Tmax, where Tmax is determined by the
TMS decorrelation factor for the longest spacing, otherwise for a large array we will never reach the desired compression factor!!

```
dphi = np.sqrt(6.  * (1. - decorrelation))
```
is missing a factor of ```pi**2``` - ie. must be ```6. / pi**2```, better yet use 
```
        asincapprox = lambda x: np.sqrt(3)/np.pi*np.arccos(x)
        dphi = asincapprox(decorrelation)
```

**Comment by sjperkins** [comment]
> This here has been the source of 2 days worth of heartache - why is this fixed to a constant value for all arrays????

I think there's tension between the need to chunk data because it's too big to fit in memory and the need to select a time range here.

Note that the 10,000 constant is used as a guideline -- if you check the log it will usually chunk fewer rows than this because it avoids splitting a timestep over two chunks.

We can discuss more about how to select a greater time range -- at the moment, the link between row and number of times in the chunk is ill-defined.

**Comment by bennahugo** [comment]
What I'm thinking of is scan through the times to find the necessary number of times (and therefore number of rows) that is required in a chunk to ensure requested decorrelation tolerances are met prior to starting. As already mentioned this is known based only on the baseline lengths, so can be computed without looking at the data or computing the actual mapping. If the row numbers suppled is less than this provide a prompt asking to proceed regardless (e.g. the user is in a memory constrained environment). An additional option to disable this interaction for pipelined environments is optional.

**Comment by sjperkins** [comment]
> 
> 
> What I'm thinking of is scan through the times to find the necessary number of times (and therefore number of rows) that is required in a chunk to ensure requested decorrelation tolerances are met prior to starting. As already mentioned this is known based only on the baseline lengths, so can be computed without looking at the data or computing the actual mapping. If the row numbers suppled is less than this provide a prompt asking to proceed regardless (e.g. the user is in a memory constrained environment). An additional option to disable this interaction for pipelined environments is optional.

Yes, I think this is doable, there's some existing infrastructure in xova to do this at present. I will look at extending it.

**Comment by sjperkins** [comment]
In fact, I think this is probably already supported:

https://github.com/ska-sa/xova/blob/7d802ca0545aee9ddda68df7d8e76ac5da2c2a6d/xova/apps/xova/chunking.py#L62-L80

but the necessary `time_bin_secs` parameter is probably not passed in in the BDA case. Thus, we should derive some sensible value from the data. Do you have an equation for doing so? I can contribute with some dask here.

**Comment by bennahugo** [comment]
Hi Simon. It is provided in the equations I specified in the memo which we
already discussed - need not be dask, but it could be. It requires in
addition to that a prefix scan operation on the time and interval column
per baseline. Once you know the decorrelation length you can work out the
maximum time  per baseline you can integrate for.

**Comment by sjperkins** [comment]
Just putting @bennahugo 's maths for this into the issue:

![2021-08-19_092453](https://user-images.githubusercontent.com/3530212/145005008-8fbe2722-86b5-46b3-8aeb-459a9ecc5a22.jpg)
